### PR TITLE
bpo-36818: Add PyInterpreterState.runtime field.

### PR DESCRIPTION
**Pull request by ericsnowcurrently**
https://bugs.python.org/issue36818

### CODE REVIEW

File: Include/internal/pycore_object.h [hunk 1]
```diff
@@ -38,11 +38,12 @@ static inline void _PyObject_GC_TRACK_impl(const char *filename, int lineno,
                           "object is in generation which is garbage collected",
                           filename, lineno, "_PyObject_GC_TRACK");
 
-    PyGC_Head *last = (PyGC_Head*)(_PyRuntime.gc.generation0->_gc_prev);
+    _PyRuntimeState *runtime = _PyRuntimeState_Get();
+    PyGC_Head *last = (PyGC_Head*)(runtime->gc.generation0->_gc_prev);
```
Comment (vstinner) [inline Include/internal/pycore_object.h]: IMHO it's wrong to put the gc structure into the runtime: it should be moved to PyInterpreterState instead, and so this function should get the current interpreter, not the runtime.

File: Include/internal/pycore_pystate.h [hunk 2]
```diff
@@ -131,11 +135,13 @@ struct _gilstate_runtime_state {
 };
 
 /* hook for PyEval_GetFrame(), requested for Psyco */
-#define _PyThreadState_GetFrame _PyRuntime.gilstate.getframe
+#define _PyThreadState_GetFrame \
+    _PyRuntimeState_Get()->gilstate.getframe
```
Comment (vstinner) [inline Include/internal/pycore_pystate.h]: IMHO gilstate should be moved into PyInterpreterState.

**Review by vstinner — CHANGES_REQUESTED** [review]
I disagree with the overall approach. Maybe I misunderstood you.
https://bugs.python.org/issue36818#msg342043

File: Include/internal/pycore_object.h (hunk 1 above)
Comment (ericsnowcurrently) [reply Include/internal/pycore_object.h]: Agreed. :)  See gh-13219.

File: Include/internal/pycore_pystate.h (hunk 2 above)
Comment (ericsnowcurrently) [reply Include/internal/pycore_pystate.h]: Also agreed.

**Comment by vstinner** [comment]
Eric is working on a different approach: https://bugs.python.org/issue36854

**Comment by ericsnowcurrently** [comment]
The GC stuff is orthogonal to adding `PyInterpreterState.runtime`.  I still plan on merging this after resolving merge conflicts and any feedback.

**Comment by ericsnowcurrently** [comment]
FYI, I've trimmed the PR down.  The changed I've removed can wait until later.

**Pull request merged**